rustls/src/server/tls13.rs

```diff
@@ -717,6 +854,28 @@ mod client_hello {
     }
 }
 
+struct ExpectAndSkipRejectedEarlyData {
+    skip_data_left: usize,
+    next: Box<hs::ExpectClientHello>,
+}
+
+impl State<ServerConnectionData> for ExpectAndSkipRejectedEarlyData {
+    fn handle(mut self: Box<Self>, cx: &mut ServerContext<'_>, m: Message) -> hs::NextStateOrError {
+        /* "The server then ignores early data by skipping all records with an external
```

Comment:
After reading the commit message, comments and code, it's still a little unclear to me how we avoid acidentally dropping traffic-encrypted application data. Are early data records separated from traffic by other handshake messages? I feel like some context is missing and/or the exact mechanism is quite spread out over the `RecordLayer`/state machine that it's pretty hard to follow what's going on exactly.

Reply:
> Are early data records separated from traffic by other handshake messages?

Yes, the EndOfEarlyData handshake message.

The multiple mechanisms doing the same thing (which is: skipping over early data messages we can't or don't want to process) reflect that there are two cases that happens in:

1. we (the server) are happy to proceed with a full handshake - in this case we can't decrypt the early data messages and have to do trial decryption until we find a message encrypted with the handshake keys -- this is the case handled by the record_layer.rs and conn.rs changes
2. we not happy to proceed with the handshake, and send a HelloRetryRequest: this means the next client message (after the early data messages) will be a plaintext ClientHello: ExpectAndSkipRejectedEarlyData  deals with that case

Ideally we could move (1) into the state machine, but I can't see a good way of doing that without also moving decryption, handshake reassembly, alert handling, etc into each state. That itself could clear up the state held in record_layer.rs, and version-specific decisions made in common conn.rs code.